File: libs/workflow_runtime/src/orchestrator.cpp
// CONV-33 轮2 实装 — workflow/orchestrator.py 游标语义（route A4）。
//
// headless 步骤游标：只推进游标、绝不持久化（docstring 自认非权威；
// 权威 = service.home_workflow_steps，audit #847-2）。状态推断全权委托
// service.infer_workflow_step_status（默认 options：无目录 → 空组合 →
// evidence-only，与 Python get_catalog() 为 None 的宿主环境等价）。
//
// 消息模板逐字（orchestrator.hpp 冻结注释 = Python L92/98/107/113）；
// step_payload 形参按 D4 裁决省略（Python 接受但忽略）。
#include "pwb/workflow_runtime/orchestrator.hpp"

#include <pwb/workflow_runtime/service.hpp>

#include <stdexcept>
#include <string>
#include <utility>

namespace pwb::workflow_runtime {
namespace {

// project.resources 真值（缺段 / 非数组 ≙ 空 —— pydantic list 属性 parity）。
bool has_resources(const domain::Json& project) {
    if (!project.is_object()) {
        return false;
    }
    const auto it = project.find("resources");
    return it != project.end() && it->is_array() && !it->empty();
}

}  // namespace

WorkflowOrchestrator::WorkflowOrchestrator(domain::Json project)
    : project_(std::move(project)) {
    steps_.reserve(std::size(kStepOrder));
    for (std::string_view step : kStepOrder) {
        steps_.emplace_back(step);
    }
}

const domain::Json& WorkflowOrchestrator::project() const noexcept {
    return project_;
}

int WorkflowOrchestrator::current_step_index() const noexcept {
    return current_step_index_;
}

WorkflowStepContext WorkflowOrchestrator::get_step_context() const {
    // 越界游标回落首步（Python L58-61：0 <= i < len 不成立取 steps[0]）。
    const std::string step_id =
        (current_step_index_ >= 0 &&
         current_step_index_ < static_cast<int>(steps_.size()))
            ? steps_[static_cast<std::size_t>(current_step_index_)]
            : steps_[0];

    std::string step_name = step_id;
    for (const auto& [id, name] : kOrchestratorStepNames) {
        if (id == step_id) {
            step_name = std::string(name);
            break;
        }
    }

    const std::string status =
        infer_workflow_step_status(project_, step_id);
    // is_valid 陷阱（L65）：warning / running 也算 valid —— 唯 complete
    // 才是完成，但未完成不阻塞推进的门槛只看 evidence-valid。
    const bool is_valid = status == "complete" || status == "running" ||
                         status == "warning";

    std::vector<std::string> prerequisites;
    if (step_id != "data_check" && !has_resources(project_)) {
        prerequisites.push_back("数据资产清单不能为空");
    }

    WorkflowStepContext context;
    context.step_id = step_id;
    context.step_name = std::move(step_name);
    context.index = current_step_index_;
    context.total_steps = static_cast<int>(steps_.size());
    context.is_valid = is_valid;
    context.prerequisites = std::move(prerequisites);
    context.status = status;
    return context;
}

StepTransitionResult WorkflowOrchestrator::next_step() {
    WorkflowStepContext context = get_step_context();
    if (!context.prerequisites.empty()) {
        std::string joined;
        for (const std::string& prereq : context.prerequisites) {
            if (!joined.empty()) {
                joined += ", ";
            }
            joined += prereq;
        }
        StepTransitionResult result;
        result.success = false;
        result.message = "无法进入下一步: " + joined;
        result.step_context = std::move(context);
        return result;
    }
    if (!context.is_valid) {
        StepTransitionResult result;
        result.success = false;
        result.message = "当前步骤 [" + context.step_name +
                         "] 尚未完成，不能进入下一步";
        result.step_context = std::move(context);
        return result;
    }
    if (current_step_index_ < static_cast<int>(steps_.size()) - 1) {
        ++current_step_index_;
        WorkflowStepContext new_context = get_step_context();
        StepTransitionResult result;
        result.success = true;
        result.message = "已成功切换至第 " +
                         std::to_string(new_context.index + 1) + " 步 [" +
                         new_context.step_name + "]";
        result.step_context = std::move(new_context);
        return result;
    }
    StepTransitionResult result;
    result.success = true;
    result.message = "工作流已全部完成";
    result.step_context = std::move(context);
    return result;
}

}  // namespace pwb::workflow_runtime

File: legacy/python_reference/product/paleo_workbench/workflow/orchestrator.py
"""WorkflowOrchestrator: headless workflow step-transition helper.

Legacy helper kept for tests/scripting: authoritative step state inference
and persistence live in :func:`paleo_workbench.workflow.service.home_workflow_steps`
(single source of truth). This orchestrator only advances a cursor; it never
persists state, so production UI must not rely on it (audit #847-2)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from paleo_workbench.project.models import ProjectDocument, ProjectMeta
from paleo_workbench.workflow.service import infer_workflow_step_status, STEP_ORDER


@dataclass
class WorkflowStepContext:
    """Read-only context for the active workflow step."""

    step_id: str
    step_name: str
    index: int
    total_steps: int
    is_valid: bool
    prerequisites: list[str]
    status: str


@dataclass
class StepTransitionResult:
    """Result payload returned by orchestrator.next_step()."""

    success: bool
    message: str
    step_context: WorkflowStepContext


class WorkflowOrchestrator:
    """Headless step-cursor helper (see module docstring; state saving claim removed)."""

    STEP_NAMES = {
        "data_check": "数据校验",
        "factor_map": "单因素图编制",
        "prediction": "地震相预测",
        "map_compile": "古地理图编绘",
        "qc": "质量检查",
        "export": "成果导出",
    }

    def __init__(self, project: ProjectDocument | None = None) -> None:
        self.project = project or ProjectDocument(meta=ProjectMeta(name="Default Project"))
        self.current_step_index = 0
        self.steps = list(STEP_ORDER)

    def get_step_context(self) -> WorkflowStepContext:
        """Deep interface method 1/2: get read-only context of the current active step."""
        if 0 <= self.current_step_index < len(self.steps):
            step_id = self.steps[self.current_step_index]
        else:
            step_id = self.steps[0]

        step_name = self.STEP_NAMES.get(step_id, step_id)
        status = infer_workflow_step_status(self.project, step_id)
        is_valid = status in {"complete", "running", "warning"}

        prereqs = []
        if step_id != "data_check" and not self.project.resources:
            prereqs.append("数据资产清单不能为空")

        return WorkflowStepContext(
            step_id=step_id,
            step_name=step_name,
            index=self.current_step_index,
            total_steps=len(self.steps),
            is_valid=is_valid,
            prerequisites=prereqs,
            status=status,
        )

    def next_step(self, step_payload: dict[str, Any] | None = None) -> StepTransitionResult:
        """Deep interface method 2/2: advance to the next step if prerequisites are satisfied.

        The current step must be evidence-valid (``is_valid``) before
        advancing — an empty project must never walk the whole strip to
        "已完成" without producing anything (audit #847-2).
        """
        ctx = self.get_step_context()
        if ctx.prerequisites:
            return StepTransitionResult(
                success=False,
                message=f"无法进入下一步: {', '.join(ctx.prerequisites)}",
                step_context=ctx,
            )
        if not ctx.is_valid:
            return StepTransitionResult(
                success=False,
                message=f"当前步骤 [{ctx.step_name}] 尚未完成，不能进入下一步",
                step_context=ctx,
            )

        if self.current_step_index < len(self.steps) - 1:
            self.current_step_index += 1
            new_ctx = self.get_step_context()
            return StepTransitionResult(
                success=True,
                message=f"已成功切换至第 {new_ctx.index + 1} 步 [{new_ctx.step_name}]",
                step_context=new_ctx,
            )

        return StepTransitionResult(
            success=True,
            message="工作流已全部完成",
            step_context=ctx,
        )
